Add recipe and diagnostics for Schlund et al., JGR: Biogeosciences, 2020

schlunma opened the pull request:
This PR adds the actual diagnostics and recipes necessary for the analysis of Schlund et al., JGR: Biogeosciences, 2020.

I'm terribly sorry for this monster PR - I started the development a long time ago and could only upload it to the public repository when my paper was accepted (a week ago, unfortunately it has not been published online yet). Since I used this exact code to do the analysis I don't think that a scientific review is necessary. I also do not ask you to do a full review since these are way to many lines of code. Maybe you can just have a look at the documentation and run the recipes?

Since I was not able to include the correct citation in this PR (no DOI available yet) I will open an issue to remind myself of doing that. It would be really nice to include this in `v2.1` so that the code availability section of the paper can be adapted accordingly.

Thanks for your patience :smile: 

* * *

**Tasks**

-   [x] [Create an issue](https://github.com/ESMValGroup/ESMValTool/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
-   [x] Give this pull request a descriptive title that can be used as a one line summary in a changelog
-   [x] Make sure your code is composed of functions of no more than 50 lines and uses meaningful names for variables
-   [x] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [ ] Preferably Codacy code quality checks pass, however a few remaining hard to solve Codacy issues are still acceptable. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.
-   [x] Please use `yamllint` to check that your YAML files do not contain mistakes
-   [x] (Only if really necessary) Add any additional dependencies needed for the diagnostic script to setup.py, esmvaltool/install/R/r_requirements.txt or esmvaltool/install/Julia/Project.toml (depending on the language of your script) and also to package/meta.yaml for conda dependencies (includes Python and others, but not R/Julia)
-   [x] If new dependencies are introduced, check that the license is compatible with [Apache2.0](https://github.com/ESMValGroup/ESMValTool/blob/master/LICENSE)

New recipe/diagnostic

-   [x] Add documentation for the recipe to the `doc/sphinx/source/recipes` folder and add a new entry to index.rst
-   [x] Add provenance information

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

Closes #1844.

valeriupredoi commented:
@schlunma 16k added lines and 14 deletions - phew, am relieved, 14 deletions only - piece of cake to review :grin: Dude, can you maybe break this down into smaller chunks - there are a lot of documentation (rst) files that can definitely go in a separate PR, together with the png pictures; we can deal with the python code in one PR I am guessing, albeit a big PR, but at least someone else can look at the documentation (maybe someone who actually knows the science?). Cheers :beer:

schlunma commented:
Cheers V., thanks for taking a look into this! As I said, I'm really really sorry for this monster PR!! Unfortunately, separating the doc from the actual code is not possible since most of the doc is created using `.. automodule::`.

As I already said, I really don't think that a scientific review is necessary since this code went basically through full peer review!

(P.S. the failing test is probably not related to this PR: `Exited with code exit status 137`. Do we have an issue for this?)

valeriupredoi commented:
aah right, completely forgot about the review from the paper (congrats on the paper, bro!). OK I will start the code review tomorrow, I can do this all - can take the burden off @bouweandela since he's reviewed a lot elsewhere so far anyway. `Exited with code exit status 137` is the CI machine running out of memory since the free account is limited to 4G only - no worries about it, try triggering the Github Action tbuild test on `push` on this PR if you are afraid the conda build will not work due to some dependency that's introduced here :beer:

axel-lauer commented on doc/sphinx/source/api/esmvaltool.rst:
```diff
@@ -1,11 +1,14 @@
 
-ESMValTool is mostly used as a commandline tool. However, it is also possibly to use (parts of) EsmValTool as a
-library. This section documents the public API of ESMValTool.
+ESMValTool is mostly used as a commandline tool. However, it is also possibly
```
```suggestion
ESMValTool is mostly used as a commandline tool. However, it is also possible
```

axel-lauer commented on doc/sphinx/source/api/esmvaltool.rst:
```diff
@@ -1,11 +1,14 @@
-ESMValTool is mostly used as a commandline tool. However, it is also possibly to use (parts of) EsmValTool as a
-library. This section documents the public API of ESMValTool.
+ESMValTool is mostly used as a commandline tool. However, it is also possibly
+to use (parts of) EsmValTool as a library. This section documents the public
```
```suggestion
to use (parts of) ESMValTool as a library. This section documents the public
```

axel-lauer commented:
Tests of recipe_schlund20jgr_gpp_abs_rcp85.yml are still running, but the ones of recipe_schlund20jgr_gpp_change_1pct.yml and recipe_schlund20jgr_gpp_change_rcp85.yml finished successfully. Here a few things I noticed so far:

* The image files in doc/sphinx/source/recipes/figures/schlund20jgr/ have a quite high resolution and large file size. I think it would be good to reduce that a little bit so the pdf of the user's guide does not get unnecessarily big.
* Within the output directory "plot", quite some subdirectories named "preprocess*" are created. Those are all empty. Is that on purpose?
* The two recipes tested so far do not provide any provenance output for the plots, only for the nc files in work.

schlunma commented:
Thanks for testing Axel!

>     * The image files in doc/sphinx/source/recipes/figures/schlund20jgr/ have a quite high resolution and large file size. I think it would be good to reduce that a little bit so the pdf of the user's guide does not get unnecessarily big.

I will re-upload them with reduced size.

>     * Within the output directory "plot", quite some subdirectories named "preprocess*" are created. Those are all empty. Is that on purpose?

That is automatically done by ESMValTool (which creates the subdirectories `plot`, `run`, and `work` for every diagnostic). Since the `preprocess` diagnostic does not write plots, the `plot` directory stays empty.

>     * The two recipes tested so far do not provide any provenance output for the plots, only for the nc files in work.

For every plot file I created an `.nc` file. The provenance output is attached to that (the plot is attached to that with attribute `plot_file`). When I wrote these scripts this was the "official" way to do this (at least it was done like this by the example python diagnostic), but this is also changing now (see discussion in #1827). I really don't want to change that since it took me 2 days to add the provenance tracking to all diagnostics in the first place...

schlunma commented:
> OK I promise I will start revieweing tomorrow (really can't start it now at 5pm grin ) but before I do that - so we make a smooth transition to iris3 - have a look at [ESMValGroup/ESMValCore#819](https://github.com/ESMValGroup/ESMValCore/pull/819) and specifically at https://scitools-iris.readthedocs.io/en/v3.0.0rc0/whatsnew/3.0.html#features (stuff you've already reviewed anyway) and was wondering if there's anything in all this code that may be an issue when switching to iris3. And I promise I'm not stalling the review grin

Thanks V.! I just checked the link; I'm not using any of the functions listed in the incompatible changes/deprecations. So (in theory) the code should run in iris3 :smile:

bouweandela commented:
> I will re-upload them with reduced size.

It might be nice to rebase, because then the commits containing the large files will not end up in the master branch, so they will be garbage collected at some point and the repo will stay small.

schlunma commented:
> > I will re-upload them with reduced size.
> 
> It might be nice to rebase, because then the commits containing the large files will not end up in the master branch, so they will be garbage collected at some point and the repo will stay small.

Sure, can do that. Does the squash commit not prevent that?

bouweandela commented:
> Does the squash commit not prevent that?

Good point, of course it does! No need to rebase then.

valeriupredoi commented:
was about to say about the squash&merge. OK I am starting to read the Python code now. Next time you should provide me with a Kindle reader @schlunma - gonna be a long read :grin:

axel-lauer reviewed (approved):
The last test just finished, looks all fine to me.

bouweandela commented on esmvaltool/recipes/schlund20jgr/recipe_schlund20jgr_gpp_change_rcp85.yml:
```diff
@@ -0,0 +1,947 @@
+    - {var_type: null}
+  imputation_strategy: mean
+  mlr_model_type: gbr_sklearn
+  n_jobs: 24
```
This looks like it specifies the number of processors to use. It might be better to set this to `null` in the recipe and then use the number of processors available in the machine, as that is more likely to work reasonably well for people using a different machine.

bouweandela replied:
E.g. using this function: https://psutil.readthedocs.io/en/latest/#psutil.cpu_count

schlunma replied:
Parallel processing in `sklearn` and in my custom functions is implemented with [`joblib.Parallel`](https://joblib.readthedocs.io/en/latest/generated/joblib.Parallel.html), which allows to specify `n_jobs=-1` to set it to the number of processors. Should I change it to that? `n_jobs=null` is identical to `n_jobs=1`.

bouweandela replied:
In that case I think it would be good to set it to `-1` in the recipe and to add this information to the documentation.

bouweandela commented on esmvaltool/recipes/schlund20jgr/recipe_schlund20jgr_gpp_change_rcp85.yml:
```diff
@@ -0,0 +1,947 @@
+    - crescendo
+    - usmile
+
+  domains:
+    - global
+
+  realms:
+    - land
+
+  themes:
+    - ML
+    - EC
+    - carbon
```
I'm not sure if it works if you specify this here, typically `themes` and `realms` are specified in the diagnostic sections and `domains` is something that can be specified in the provenance record coming from the diagnostic. See also the documentation on recording provenance: https://docs.esmvaltool.org/en/latest/community/diagnostic.html#recording-provenance

schlunma replied:
Ah, I didn't know that. Would it make sense to implement that? I think we should be very flexible here (allow setting all provenance options in the recipe, diagnostic section and diagnostic itself). In this case it would make sense to set these option in the recipe for all diagnostic. Setting them to fixed values in the diagnostic script itself makes no sense since it works on arbitrary input variables (type and shapes).

bouweandela replied:
This is how it's currently implemented in the ESMValCore, so for now I think it is best to implement that here. Maybe you could open an issue in the ESMValCore repository to propose to make this more flexible?

schlunma replied:
So you mean adding this to every diagnostic? This won't look nice, but I can do it. I will open an issue about that.

bouweandela replied:
> So you mean adding this to every diagnostic?

Yes, because then it works, see e.g. examples/recipe_python.yml or recipe_perfmetrics_CMIP5.yml for how it's done there

valeriupredoi commented on esmvaltool/diag_scripts/mlr/custom_sklearn.py:
```diff
@@ -0,0 +1,771 @@
+        else:
+            msg = '%s' % (', '.join('%s=%s' % (k, v)
+                                    for k, v in parameters.items()))
+        print("[CV] %s %s" % (msg, (64 - len(msg)) * '.'))
```
logger instead of print?

schlunma replied:
I want to stay as close to the original `sklearn` implementation as possible, so I'd like to keep that.

valeriupredoi replied:
ok

valeriupredoi commented on esmvaltool/diag_scripts/mlr/custom_sklearn.py:
```diff
@@ -0,0 +1,771 @@
+        """Additional estimator tags."""
+        more_tags = super()._more_tags()
+        more_tags['allow_nan'] = True
+        return more_tags
```
probably the best way to review this, rather than a line by line review, which is tedious and error-prone for everyone but you and the sklearn module devvs would be to summarize in a few bullet points what custom changes you made to the scipy module and add it to the documentation (something like technical stuff, if it's in your paper just lift it from there)

schlunma replied:
I don't think further doc is necessary. The changes made are either in the name of the function, e.g. `_score_weighted` or `_fit_and_score_weighted` instead of `_score` and `_fit_and_score` or listed in the docstring, e.g. `"""Expand :meth:`fit` to accept kwargs."""` or `"""Expand :meth:`_fit` to accept kwargs."""`. Since this is rather technical stuff this is not in the paper :grimacing:

valeriupredoi replied:
okay - have you thought about contributing directly to `scipy` and see if these things can be included? I am a bit apprehensive to have a lot of code lifted from a library that evolves with time and dropped+changed locally (wrt `scipy`, not locally to us) - I am 100% sure of what you did is right but I don't want to have a wise guy come over to us in a year or so and "find bugs" and say we made a Chinese copy. If these changes were documented to the very least in a document we'd be able to say look we made these changes, tested them and it fit our purpose

schlunma replied:
I fully agree with you - it would be optimal to have this in the `sklearn` itself, but I really do not have the time to do this right now, sorry :disappointed: I can expand the tests (there are already quite some) in a future PR to ensure that these custom changes are working if necessary, but I cannot provide any more documentation on this than already given by the docstring and the comments in the code, and moving them to a different document is not useful in my opinion.

valeriupredoi replied:
no, of course not now, that process is very lengthy - ok meet me middle ground then, how about a few lines worth of paragraph in a footnote in the documentation saying that you took these funcs and expanded them to fit the purposes for the desired functionality and you are open to contribute these features to the scipy's library - and mention that tests cover these as well and will increase the test coverage in the future even more. This should cover our arses and you don't have to do much more work too (well, apart from improving the test coverage in a future PR, after you finish writing your thesis :grin: )

schlunma replied:
Agreed :+1: I added a note to documentation of the module, please check if you agree with it!

valeriupredoi commented on esmvaltool/diag_scripts/mlr/main.py:
```diff
@@ -0,0 +1,349 @@
+if __name__ == '__main__':
+    mlr.ignore_warnings()
+    with run_diagnostic() as config:
+        main(config)
```
ON, stopping here for now

schlunma replied:
Cheers, thanks for the review so far!

valeriupredoi reviewed (commented):
Stopped right before `mmm.py` - will resume later :+1:

schlunma commented on esmvaltool/diag_scripts/mlr/__init__.py:
```diff
@@ -0,0 +1,837 @@
+                          long_name='Land mask (1: land, 0: sea)',
+                          units='no_unit',
+                          dim_coords_and_dims=[(lat_coord, 0), (lon_coord, 1)])
+    return cube
```
I only use this mask in this file where I take care that the final output is on 0...360:

https://github.com/ESMValGroup/ESMValTool/blob/c2f6850783a1b515a8e21a2b6d4a7699d7d49957/esmvaltool/diag_scripts/mlr/__init__.py#L584-L585

Maybe we could add it to ESMValCore, but I think this is not necessary here.

valeriupredoi replied:
nah just a though for future work in Core, all good here, enough code to look at anyway :grin:

schlunma commented on esmvaltool/diag_scripts/mlr/__init__.py:
```diff
@@ -0,0 +1,837 @@
+        raise ValueError(
+            f"Expected identical shapes for 'x_data' and 'y_data', got "
+            f"{x_data.shape} and {y_data.shape}, respectively")
+    aux_coord = iris.coords.AuxCoord(x_data, **x_kwargs)
```
Yes!

```python
>>> import iris
>>> auxcoord = iris.coords.AuxCoord(0.0)
>>> print(auxcoord)
AuxCoord(array([0.]), standard_name=None, units=Unit('1'))
```

If `iris` allows this, we should allow it, too.

schlunma commented on esmvaltool/diag_scripts/mlr/__init__.py:
```diff
@@ -0,0 +1,837 @@
+    """
+    if units.is_time_reference():
+        units = Unit(units.symbol.split()[0])
+    if not units.is_time():
```
This is a good question, I honestly do not know, but I haven't seen other cases.

valeriupredoi replied:
aha so it's a 'if it ain't broken don't fix it' thing...ok, if it works here, let it be then

valeriupredoi replied:
note that eg `cf_units.Unit("kg s-1").symbol.split()[0]` will return `kg.s-1` so am not sure why you doing the splitting and indexing though...meh, let's not `split` hairs, if this works for you OK

schlunma replied:
It's only meant for time units (if used with other units a `ValueError` is raised). The use of this function is e.g. to convert from `days since 1850-01-01` to `days`.

schlunma commented on esmvaltool/diag_scripts/mlr/__init__.py:
```diff
@@ -0,0 +1,837 @@
+        for lon_idx in range(lon_coord.shape[0]):
+            lat_bounds = lat_coord.bounds[lat_idx]
+            lon_bounds = lon_coord.bounds[lon_idx]
+            submask = ne_land_mask_cube.intersection(latitude=lat_bounds,
```
As far as I understand this issue, no. `ne_land_mask_cube` has no coordinate boundaries.

schlunma commented on esmvaltool/diag_scripts/mlr/__init__.py:
```diff
@@ -0,0 +1,837 @@
+            submask = ne_land_mask_cube.intersection(latitude=lat_bounds,
+                                                     longitude=lon_bounds)
+            land_fraction[lat_idx, lon_idx] = (submask.data.sum() /
+                                               submask.data.size)
```
As I already noted this is slow, but still reasonable fast if not called often (which I don't do). I don't think it's necessary to put more work into this.

valeriupredoi replied:
OK up to you, at least you seen the thing run, I have no feel for how slow or fast it is, it just *looks* like it's slow on its own. Resolved then

schlunma commented on esmvaltool/diag_scripts/mlr/__init__.py:
```diff
@@ -0,0 +1,837 @@
+        new_data = cube.data
+        if not cube.attributes.get('squared'):
+            new_data **= 2
+        squared_error_cube.data += new_data
```
I could, but since large parts of this code use `sklearn` (which can't handle `dask` at the moment) I don't think that this would change anything. The `sklearn` parts need all data realized anyway :smile:

valeriupredoi replied:
ah crap, ok then :+1:

schlunma commented on esmvaltool/diag_scripts/mlr/__init__.py:
```diff
@@ -0,0 +1,837 @@
+
+def square_root_metadata(cube):
+    """Take the square root of the cube metadata.
+
```
That's exactly what I'm doing :smile: This code either removes `squared (if possible) or adds `root` to the metadata (and applies a root to the units, of course).

valeriupredoi replied:
hmm maybe I misinterpreted the squaring, nevermind

bouweandela commented:
You might want to have a closer look at the documentation by clicking the 'Details' link behind the readthedocs build below or building the documentation locally. I noticed several problems, especially in the API documentation:
- links are not correctly formatted in several places
- the data type seems to be concatenated to the name of the setting that can be used in the diagnostic section of the recipe
- there are many functions documented that may be of little interest to the user (and the documentation is hardly informative, for example [here](https://esmvaltool--1860.org.readthedocs.build/en/1860/api/esmvaltool.diag_scripts.mlr/evaluate_residuals.html), the recipe options are useful, but the rest (documentation of classes and functions in the module) does not really help much. I do not see anyone re-using this as a Python module based on the information provided here.
- maybe you need to pin autodocsumm < 1.2, because I can see that the documentation is suffering from this issue: https://github.com/Chilipp/autodocsumm/issues/33

schlunma commented:
>     * links are not correctly formatted in several places

This does not happen in my local build. Any ideas why?

>     * the data type seems to be concatenated to the name of the setting that can be used in the diagnostic section of the recipe

This does also not happen in my local build...

>     * there are many functions documented that may be of little interest to the user (and the documentation is hardly informative), for example [here](https://esmvaltool--1860.org.readthedocs.build/en/1860/api/esmvaltool.diag_scripts.mlr/evaluate_residuals.html), the recipe options are useful, but the rest (documentation of classes and functions in the module) does not really help much. I do not see anyone re-using this as a Python module based on the information provided here.

I agree, but

1. copy/pasting these recipe options to an `rst` file is error prone and not nice (I wrote some diagnostics + doc where I forgot some options/had wrong options), so by doing it like this we avoid duplication.
2. It's in the API section of the diagnostic, I don't think it does any harm there.

>     * maybe you need to pin autodocsumm < 1.2, because I can see that the documentation is suffering from this issue: [Chilipp/autodocsumm#33](https://github.com/Chilipp/autodocsumm/issues/33)

Yes, I already noticed that will try that!

bouweandela commented:
> This does not happen in my local build. Any ideas why?

Probably because you're using a different version of the software than readthedocs.

schlunma commented:
So that means readthedocs doesn't use the information given e.g here?

https://github.com/ESMValGroup/ESMValTool/blob/d4f93d25d3643f647ceae5495ef4c7f5f3f1980e/setup.py#L79

Does that mean I have to do a push every time I change something and wait for readthedocs when I want to change the documentation? That's not really a nice way of developing.

bouweandela commented:
> I agree, but
>
> 1    copy/pasting these recipe options to an rst file is error prone and not nice (I wrote some diagnostics + doc where I forgot some options/had wrong options), so by doing it like this we avoid duplication.
> 2    It's in the API section of the diagnostic, I don't think it does any harm there.

There is no need to copy/paste, you can specify exactly what should be documented in the .rst files by using [autodoc directives](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html), to override the default from https://github.com/ESMValGroup/ESMValTool/blob/c2f6850783a1b515a8e21a2b6d4a7699d7d49957/doc/sphinx/source/conf.py#L65-L71

For more fine grained control, you can also add an `__all__ = ['interesting_function_1', 'interesting_function_2']` to the modules you're documenting, only stuff from `__all__` will appear in the documentation.

bouweandela commented:
> So that means readthedocs doesn't use the information given e.g here?

readthedocs uses [`doc/sphinx/source/requirements.txt`](https://github.com/ESMValGroup/ESMValTool/blob/add_schlund20jgr/doc/sphinx/source/requirements.txt) to create a virtual environment in which it build the documentation.

schlunma commented:
> > So that means readthedocs doesn't use the information given e.g here?
> 
> readthedocs uses [`doc/sphinx/source/requirements.txt`](https://github.com/ESMValGroup/ESMValTool/blob/add_schlund20jgr/doc/sphinx/source/requirements.txt) to create a virtual environment in which it build the documentation.

Wouldn't it make sense then that these two files are somehow similar? E.g. `setup.py` includes `sphinx_rtd_theme`, `requirements.txt` doesn't. Might this cause issues?

schlunma commented:
I could solve the problem with the links and removed all unnecessary information from the documentation of the diagnostics.

Pinning `autodocsum` didn't help since I was using `0.2.1` anyway: 

```
...
autodocsumm               0.2.1                    pypi_0    pypi
...
```

I also have no idea about the weird formatting of recipe options section, I couldn't find anything relevant on Google. In my local build it looks like this:

![mlrdoc](https://user-images.githubusercontent.com/32543114/96280360-29964a80-0fd8-11eb-878a-465192fd3180.png)

bouweandela commented:
> Pinning autodocsum didn't help since I was using 0.2.1 anyway:

The version numbers in the referenced issue are wrong, you would need to pin it to <0.2

schlunma commented:
Alright, I made some more progress:

@bouweandela I fixed all issues now by pinning `autodocsum` to `<0.2.0` and removing spaces between colons to fix the formatting of the configuration options(turns out the problem was a space between the colons...).

@valeriupredoi I added the comment about the custom `sklearn` functions, see [here](https://esmvaltool--1860.org.readthedocs.build/en/1860/api/esmvaltool.diag_scripts.mlr/custom_sklearn.html).

Any more comments from your side?

valeriupredoi reviewed (approved):
looks good to me and helluva lot of work, well done @schlunma

bouweandela commented on esmvaltool/recipes/schlund20jgr/recipe_schlund20jgr_gpp_change_1pct.yml:
```diff
@@ -0,0 +1,381 @@
+# YAML anchors
+
+PROVENANCE_ANCHOR: &provenance
+  domains:
+    - global
```
```suggestion
```
This doesn't work, domains can only be specified in the provenance coming from the diagnostic, see also my earlier comment: https://github.com/ESMValGroup/ESMValTool/pull/1860#pullrequestreview-510390248. Probably best to remove it everywhere, since it doesn't work anyway if you add it like this.

bouweandela commented:
Thanks, the documentation looks much better now, though it's still not exactly an easy read.

Apart from the minor comment that the provenance is still not quite right, I think this can be merged soon. I'm not going to look at the code since @valeriupredoi already did that.

I'm running the conda build locally now, just to make sure it's still working.
